core/primitives/src/version.rs

```diff
@@ -166,6 +166,14 @@ lazy_static! {
         nightly_protocol_features_to_version_mapping
     };
 }
+#[cfg(all(feature = "nightly_protocol", feature = "protocol_feature_block_header_v3"))]
+lazy_static! {
+    pub static ref BLOCK_HEADER_V3_VERSION: ProtocolVersion =
+        *PROTOCOL_FEATURES_TO_VERSION_MAPPING.get(&ProtocolFeature::BlockHeaderV3).unwrap();
+}
+
+#[cfg(not(feature = "protocol_feature_block_header_v3"))]
+pub const BLOCK_HEADER_V3_VERSION: &'static ProtocolVersion = &u32::MAX;
```

Comment:
What is this for?

Reply:
`BLOCK_HEADER_V3_VERSION` is used to select runtime behaviour based on the protocol version. For example it is used [here](https://github.com/near/nearcore/pull/4115/files#diff-bf94debb856e8e44f62c9316b1e0c96d25e57610e3cc02481abe3737c5d2b82bR1147) to decide what version of the chunk header to produce.

Needing to choose logic based on the protocol version (as opposed to compile-time features) is also done elsewhere. For example, it is done [here](https://github.com/near/nearcore/blob/master/core/primitives/src/block_header.rs#L382) to decide which version of the block header to produce. However, in that code, there is an explicit runtime lookup of the `PROTOCOL_FEATURES_TO_VERSION_MAPPING`, which seems unnecessary to me because it is a constant. So I have introduced `BLOCK_HEADER_V3_VERSION` which essentially caches the version lookup.

To make the code compile I also need `BLOCK_HEADER_V3_VERSION` to be defined when the `protocol_feature_block_header_v3` is not enabled, so I inserted a value which says that the block header v3 version is impossibly far in the future. I could probably eliminate the need to have this constant defined when the feature is not enabled, but it might mean making some of the `if` statements where it would be used less clear because additional compile-time feature gating would be needed.

Reply:
My question is why this is done the same way as how other protocol upgrades (you can check where `ProtocolFeatures` variants are used and how they used to determine the version). Is there something specific about this change?

Reply:
The main difference is there are multiple possible behaviours, i.e. it is not binary. We need to select either version 1, 2, or 3 depending on the protocol version for both block headers and chunk headers, and this is implemented as `if (cond1) ... else if (cond2) ... else`, where the conditions are checking the current protocol version against known points where the logic changed.

As I mentioned in my previous comment, the logic for block header was already done in this fashion, but used runtime lookup on `PROTOCOL_FEATURES_TO_VERSION_MAPPING` instead of a constant. I thought using a constant was better for both readability and performance.

Reply:
I think what you did here defeats the purpose of having `ProtocolFeature` and `PROTOCOL_FEATURES_TO_VERSION_MAPPING`. If done this way, we might as well have a constant for each version to begin with.

Reply:
Yes, you are right that it does not effectively make use of those structures. But the main question for me is whether it would be better if I were to use them more directly. I don't think it does, but if you do then I will change it to match what is done with block headers.

Reply:
I think the reason why we use `ProtocolFeature` and `PROTOCOL_FEATURES_TO_VERSION_MAPPING` is to put all versions in one place so that it is easier to manage. For performance, I think it is a valid concern. @matklad do you have any suggestions here?

Reply:
If the primary concern here is runtime lookup, it can be eliminated by using a (const-evalable) flat array instead of hash map:

```rust
#[derive(Clone, Copy)]
#[repr(usize)]
pub enum Feature {
    A,
    #[cfg(feature = "b")]
    B,
    __N_Features,
}

pub const fn feature_to_version(f: Feature) -> u32 {
    const MAPPING: [u32; Feature::__N_Features as usize] = {
        let mut res = [0; Feature::__N_Features as usize];
        let map = [
            (Feature::A, 100),
            #[cfg(feature = "b")]
            (Feature::B, 200),
            ];

        let mut idx = 0;
        while idx < map.len() {
            res[map[idx].0 as usize] = map[idx].1;
            idx += 1;
        }

        let mut idx = 0;
        while idx < res.len() {
            let is_set = res[idx] != 0;
            let _assert= [()][!is_set as usize];
            idx += 1;
        }

        res
    };
    MAPPING[f as usize]
}
```

Alternatively, we might want to just make protocol version feature's discriminant?

```rust
#[derive(Clone, Copy)]
#[repr(u32)]
pub enum Feature {
    A = 100,
    #[cfg(feature = "b")]
    B = 200,
}
```

Taking a step back, I think that:

* the current API for feature checking feels awkward, and probably can be improved in terms of ergonomics, runtime preformance, and runtime semantics
* there's a large benefit in uniformity (even if it results in not the best possible code)
* we probably should look into cleaning this up in a separate PR

Reply:
Yeah I agree with what @matklad said. Let's just follow the current way of doing the protocol version check in this PR and then we can address the issue separately. Created #4171

Reply:
Sounds good @bowenwang1996 

Changed to using the mapping directly in https://github.com/near/nearcore/pull/4115/commits/a21a48b96f40be81449cfbacf4e961d590ef6240